File: core/update.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import urllib    # used for getting Launchpad page
import os        # used for some files manipulations
import datatime  # used for getting the date
import sys       # used for shutting down the script
import shutil    # used for backup

from datetime import datetime




"""------------------------------------
   ----- VARIOUS VARIABLES ------------
   ---------------------------------"""

url              = 'https://launchpad.net/elementary/+milestone/freya-beta2' # URL of the Launchpad page you want
data_file        = 'data.csv'                                                # name of you data file
data_backup_file = 'data_backup.csv'                                         # name of your backup file
temp_file        = 'temp.del'                                                # name of temp file
bugs_line        = 500                                                       # max line of url page source code where bugs count can be found
date             = datetime.utcnow().strftime("%d/%m/%y")                    # the date (UTC)




"""------------------------------------
   ----- LET'S START ------------------
   ---------------------------------"""


# Check if today's datas are here, if true -> let's end the script

if date in open(data_file).read():
    print('no update needed')
    sys.exit()


# Get Launchpad page

response = urllib.urlopen(url)
data     = response.read()
response.close()


# Write data to temp file.

ff = open(temp_file, "w")
ff.write(data)
ff.close()


# Find number of bugs

b_new           = 0
b_incomp        = 0
b_conf          = 0
b_inprog        = 0
b_triaged       = 0
b_fix_committed = 0
b_fix_released  = 0


temp = open(temp_file, "r")
for i, line in enumerate(temp):

        # New bugs.
        if 'span class="statusNEW">' in line and i < bugs_line:
                a = temp.next()
                b = a.split('<strong>')
                c = b[1].split('</strong>')
                b_new = int(c[0])

        # Incomplete bugs.
        if 'span class="statusINCOMPLETE">' in line and i < bugs_line:
                a = temp.next()
                b = a.split('<strong>')
                c = b[1].split('</strong>')
                b_incomp = int(c[0])

        # Confirmed bugs.
        if 'span class="statusCONFIRMED">' in line and i < bugs_line:
                a = temp.next()
                b = a.split('<strong>')
                c = b[1].split('</strong>')
                b_conf = int(c[0])

        # Triaged bugs.
        if 'span class="statusTRIAGED">' in line and i < bugs_line:
                a = temp.next()
                b = a.split('<strong>')
                c = b[1].split('</strong>')
                b_triaged = int(c[0])

        # In progress bugs.
        if 'span class="statusINPROGRESS">' in line and i < bugs_line:
                a = temp.next()
                b = a.split('<strong>')
                c = b[1].split('</strong>')
                b_inprog = int(c[0])

        # In progress bugs.
        if 'span class="statusFIXCOMMITTED">' in line and i < bugs_line:
                a = temp.next()
                b = a.split('<strong>')
                c = b[1].split('</strong>')
                b_fix_committed = int(c[0])

        # In progress bugs.
        if 'span class="statusFIXRELEASED">' in line and i < bugs_line:
                a = temp.next()
                b = a.split('<strong>')
                c = b[1].split('</strong>')
                b_fix_released = int(c[0])


# Delete temp file

os.remove(temp_file)


# Backing up data.csv

shutil.copy2(data_file, data_backup_file)


# Print bugs in .csv file

input = str(date) + ", " + str(b_new) + ', ' + str(b_incomp) + ', ' + str(b_conf) + ', ' + str(b_triaged)  + ', ' + str(b_inprog) + ', ' + str(b_fix_committed) + ', ' + str(b_fix_released)

with open(data_file, 'a') as file:
    file.write('\n' + input)
    print (data_file + ' updated')

# I guess it's the end

File: update.php
#!/usr/bin/php

<?php


    /*
    |--------------------------------------------------------------------------
    | Config
    |--------------------------------------------------------------------------
    */

    date_default_timezone_set("UTC");



    /*
    |--------------------------------------------------------------------------
    | Variables
    |--------------------------------------------------------------------------
    */

    $url              = 'https://launchpad.net/elementary/+milestone/freya-rc1'; // URL of the Launchpad page you want
    $data_file        = 'data.csv';                                              // name of your data file
    $temp_file        = 'temp.del';                                              // name of temp file
    $bugs_line        =  300;                                                    // max line of url page source code where bugs
    $current_date     =  date('d/m/y');                                          // the date (UTC)



    /*
    |--------------------------------------------------------------------------
    | Get bugs count
    |--------------------------------------------------------------------------
    */

    // Check if the script was runned once - best for memory cunsomption

    if(exec('grep '.escapeshellarg($current_date).' ./data.csv')) exit();


    // Get Lauchpad page

    $context = stream_context_create(array('http' => array('header'=>'Connection: close\r\n'))); // Fix performance issue
    $reponse = file_get_contents($url, false, $context);
    $reponse = implode("\n", array_slice(explode("\n", $reponse), $bugs_line));                  // Shortern String for better performance

    $array = explode("\n", $reponse);

    foreach($array as $i=>$line) {

        if(strpos($line,'class="statusNEW"') !== false && !isset($b_new))
        {

            $b_new = str_replace([' ', '<strong>', '</strong>'], '', $array[$i + 1]);
        }
        else if(strpos($line,'class="statusINCOMPLETE"') !== false && !isset($b_incomp))
        {

            $b_incomp = str_replace([' ', '<strong>', '</strong>'], '', $array[$i + 1]);
        }
        else if(strpos($line,'class="statusCONFIRMED"') !== false && !isset($b_conf))
        {

            $b_conf = str_replace([' ', '<strong>', '</strong>'], '', $array[$i + 1]);
        }
        else if(strpos($line,'class="statusTRIAGED"') !== false && !isset($b_triaged))
        {

            $b_triaged = str_replace([' ', '<strong>', '</strong>'], '', $array[$i + 1]);
        }
        else if(strpos($line,'class="statusINPROGRESS"') !== false && !isset($b_inprog))
        {

            $b_inprog = str_replace([' ', '<strong>', '</strong>'], '', $array[$i + 1]);
        }
        else if(strpos($line,'class="statusFIXCOMMITTED"') !== false && !isset($b_fix_committed))
        {

            $b_fix_committed = str_replace([' ', '<strong>', '</strong>'], '', $array[$i + 1]);
        }
        else if(strpos($line,'class="statusFIXRELEASED"') !== false && !isset($b_fix_released))
        {

            $b_fix_released = str_replace([' ', '<strong>', '</strong>'], '', $array[$i + 1]);
        }
    }

    $b_new           = isset($b_new)           ? $b_new           : 0;
    $b_incomp        = isset($b_incomp)        ? $b_incomp        : 0;
    $b_conf          = isset($b_conf)          ? $b_conf          : 0;
    $b_inprog        = isset($b_inprog)        ? $b_inprog        : 0;
    $b_triaged       = isset($b_triaged)       ? $b_triaged       : 0;
    $b_fix_committed = isset($b_fix_committed) ? $b_fix_committed : 0;
    $b_fix_released  = isset($b_fix_released)  ? $b_fix_released  : 0;



    /*
    |--------------------------------------------------------------------------
    | Saving new data
    |--------------------------------------------------------------------------
    */


    // Generate input

    $input = $current_date.', '.$b_new.', '.$b_incomp.', '.$b_conf.', '.$b_triaged .', '.$b_inprog.', '.$b_fix_committed.', '.$b_fix_released;


    // Write in data.csv

    $file = fopen($data_file, 'a');
    fwrite($file, $input."\n");
    fclose($file);


    // We're done
